packages/flutter/lib/src/material/material_state.dart

```diff
@@ -735,11 +735,21 @@ class MaterialStatePropertyAll<T> implements MaterialStateProperty<T> {
 ///
 /// Used by widgets that expose their internal state for the sake of
 /// extensions that add support for additional states. See
-/// [TextButton.statesController] for example.
+/// [TextButton] for an example.
 ///
 /// The controller's [value] is its current set of states. Listeners
 /// are notified whenever the [value] changes. The [value] should only be
 /// changed with [update]; it should not be modified directly.
+///
+/// The controller's [value] represents the set of states that a
+/// widget's visual properties, typically [MaterialStateProperty]
+/// values, are resolved against. It is _not_ the intrinsic state of
+/// the widget. The widget is responsible for ensuring that the
+/// controller's [value] tracks its intrinsic state. For example one
+/// cannot request the keyboard focus for a widget by adding
+/// [MaterialState.focused] to its controller. When the widget gains the
+/// or loses the focus it will [update] its controller's [value] and
```

Comment:
nit: it will [update] -> the widget should typically [update] it's controller's [value] to reflect the focus change visually?

, since the controller can override the visual appearance?

Reply:
You're right that the widget isn't required to update its MaterialStatesController to match its intrinsic state. There are no examples of widgets that don't do this and I don't know of a use case where it would be useful to do so, so I think it's OK to leave this strong statement in place.